### PR TITLE
Nightly Release Test POC

### PR DESCRIPTION
**Pull request by kenny-ezirim**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/conftest.py [hunk 1]
```diff
@@ -753,7 +783,7 @@ def pytest_generate_tests(metafunc):
     # Don't parametrize if there are no images to parametrize
     if not images:
         return
-
+
```
Comment (saimidu) [inline test/dlc_tests/conftest.py:786]: nit: Please revert this change that introduces extra whitespace on this line.

File: test/dlc_tests/conftest.py [hunk 2]
```diff
@@ -89,6 +94,13 @@
     "inference",
 )
 
+# Immutable constant dictionary for nightly specific image fixtures
+# Each fixture must have a set of labels
+NIGHTLY_FIXTURES = {
+    "feature_smdebug_present": {"aws_framework_installed", "smdebug_installed"},
+    "feature_smddp_present": {"aws_framework_installed", "smddp_installed"},
+}
```
Comment (saimidu) [inline test/dlc_tests/conftest.py:102]: Please also add `"feature_smmp_present"`, with label `"smmp_installed"` associated to it. I don't believe it needs `aws_framework_installed`.

File: test/dlc_tests/test_type_images.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (saimidu) [inline test/dlc_tests/test_type_images.json:1]: Please remove this from the files submitted in the PR.

File: test/dlc_tests/conftest.py (hunk 2 above)
Comment (saimidu) [reply test/dlc_tests/conftest.py:102]: Please also add `"feature_aws_framework_present"`, with label `"aws_framework_installed"` associated to it.

File: test/dlc_tests/test_type_images.json (hunk 3 above)
Comment (kenny-ezirim) [reply test/dlc_tests/test_type_images.json:1]: This is a draft PR. I will remove all the miscellaneous files when we are ready to convert the PR and before merging to master.

File: test/dlc_tests/conftest.py [hunk 4]
```diff
@@ -89,6 +93,31 @@
     "inference",
 )
 
+# Nightly image fixture dictionary, maps nightly fixtures to set of image labels
+NIGHTLY_FIXTURES = {
+    "feature_smdebug_present": {"aws_framework_installed", "smdebug_installed"},
+    "feature_smdp_present": {"aws_framework_installed", "smddp_installed"},
```
Comment (saimidu) [inline test/dlc_tests/conftest.py:99]: Please name it smdp or smddp consistently.

File: test/test_utils/imageutils.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+import re
+import boto3
+import json
+
+def get_image_account_id(image_uri):
```
Comment (saimidu) [inline test/test_utils/imageutils.py:5]: Please add 2 line spaces before the start of each function.

Comment (saimidu) [inline test/test_utils/imageutils.py:5]: We already have an equivalent function in test_utils/__init__.py.

https://github.com/aws/deep-learning-containers/blob/5f02453540ccf56862355b6864a746a2765fd4ed/test/test_utils/__init__.py#L1091-L1098

Is this meant to replace that function? Should it the other function be removed and replaced everywhere else?

File: test/test_utils/imageutils.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    Find the account ID where the image is located
+    :param image_uri: <str> ECR image URI
+    :return: <str> AWS Account ID
+    """
+    return image_uri.split(".")[0]
+
+def get_image_repository_name(image_uri):
+    repository_uri, _ = image_uri.split(":")
+    _, repository = repository_uri.split("/")
+    return repository
```
Comment (saimidu) [inline test/test_utils/imageutils.py:16]: We already have an equivalent function in test_utils/__init__.py.

https://github.com/aws/deep-learning-containers/blob/5f02453540ccf56862355b6864a746a2765fd4ed/test/test_utils/__init__.py#L258-L265

Is this meant to replace that function? Should it the other function be removed and replaced everywhere else?

File: test/test_utils/imageutils.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+def get_image_tag_name(image_uri):
+    _, tag = image_uri.split(":")
+    return tag
```
Comment (saimidu) [inline test/test_utils/imageutils.py:20]: We already have an equivalent function in test_utils/init.py.

https://github.com/aws/deep-learning-containers/blob/5f02453540ccf56862355b6864a746a2765fd4ed/test/test_utils/__init__.py#L1485-L1494

Is this meant to replace that function? Should it the other function be removed and replaced everywhere else?

File: test/test_utils/imageutils.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+def get_image_region(image_uri):
+    """
+    Find the region where the image is located
+    :param image_uri: <str> ECR image URI
+    :return: <str> AWS Region Name
+    """
+    region_pattern = r"(us(-gov)?|ap|ca|cn|eu|sa)-(central|(north|south)?(east|west)?)-\d+"
+    region_search = re.search(region_pattern, image_uri)
+    assert region_search, f"{image_uri} must have region that matches {region_pattern}"
+    return region_search.group()
```
Comment (saimidu) [inline test/test_utils/imageutils.py:31]: We already have an equivalent function in test_utils/init.py.

https://github.com/aws/deep-learning-containers/blob/5f02453540ccf56862355b6864a746a2765fd4ed/test/test_utils/__init__.py#L1101-L1111

Is this meant to replace that function? Should it the other function be removed and replaced everywhere else?

File: test/test_utils/imageutils.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    """
+    account_id = get_image_account_id(image_uri)
+    repo = get_image_repository_name(image_uri)
+    tag = get_image_tag_name(image_uri)
+    region = get_image_region(image_uri)
+    if not client:
+        client = boto3.client("ecr", region_name=region)
+
+    labels = get_image_labels_with_manifest(client, repo, tag, account_id=account_id)
+    return labels
```
Comment (saimidu) [inline test/test_utils/imageutils.py:48]: We already have an equivalent function in test_utils/init.py.

https://github.com/aws/deep-learning-containers/blob/5f02453540ccf56862355b6864a746a2765fd4ed/test/test_utils/__init__.py#L1735-L1767

Is this meant to replace that function? Should it the other function be removed and replaced everywhere else?

File: test/test_utils/imageutils.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        raise KeyError(f"imageManifest not found in ecr_client.batch_get_image response:\n{response['images']}")
+    return response["images"][0]["imageManifest"]
+
+def are_image_labels_matched(image_uri, labels):
+    image_label_collection = get_image_labels(image_uri)
+    if not image_label_collection:
+        return False
+    image_labels = image_label_collection.keys()
+    for label in labels:
+        if label not in image_labels or image_label_collection[label] != "true":
```
Comment (saimidu) [inline test/test_utils/imageutils.py:98]: You can directly use `if label not in image_label_collection` instead of `if label not in image_label_collection.keys()`.

Also, please use `image_label_collection[label].lower() != "true"` to account for any case-differences, such as "TRUE" or "True".

File: test/dlc_tests/conftest.py [hunk 11]
```diff
@@ -540,7 +569,6 @@ def pt15_and_above_only():
 def pt14_and_above_only():
     pass
 
-
```
Comment (saimidu) [inline test/dlc_tests/conftest.py:543]: Please add this back.

File: test/dlc_tests/conftest.py [hunk 12]
```diff
@@ -753,7 +781,6 @@ def pytest_generate_tests(metafunc):
     # Don't parametrize if there are no images to parametrize
     if not images:
         return
-
```
Comment (saimidu) [inline test/dlc_tests/conftest.py:756]: Please add this back.

File: test/dlc_tests/ec2/test_smdebug.py [hunk 13]
```diff
@@ -15,7 +15,7 @@
 SMDEBUG_EC2_GPU_INSTANCE_TYPE = get_ec2_instance_type(default="p3.8xlarge", processor="gpu")
 SMDEBUG_EC2_CPU_INSTANCE_TYPE = get_ec2_instance_type(default="c4.8xlarge", processor="cpu")
 
-
+@pytest.mark.usefixtures("feature_smdebug_present")
```
Comment (saimidu) [inline test/dlc_tests/ec2/test_smdebug.py:18]: Please leave 2 lines of space before the start of a function.

```suggestion

@pytest.mark.usefixtures("feature_smdebug_present")
```

File: test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py [hunk 14]
```diff
@@ -75,7 +75,6 @@ def get_transformers_version(ecr_image):
     else:
         raise LookupError("HF transformers version not found in image URI")
 
-
```
Comment (saimidu) [inline test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py:78]: Please add this back.

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smtrcomp.py [hunk 15]
```diff
@@ -166,7 +166,7 @@ def test_horovod(self, sagemaker_session, ecr_image, framework_version, instance
         captured = capsys.readouterr()
         _assert_training_compiler_invoked(captured)
 
-
+    @pytest.mark.usefixtures("feature_smdp_present")
```
Comment (saimidu) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smtrcomp.py:169]: Please leave 2 lines of space before the start of a function.

```suggestion

    @pytest.mark.usefixtures("feature_smdp_present")
```

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smtrcomp.py [hunk 16]
```diff
@@ -188,7 +188,7 @@ def test_smdp(self, sagemaker_session, ecr_image, framework_version, instance_co
         captured = capsys.readouterr()
         _assert_training_compiler_invoked(captured)
 
-
+    @pytest.mark.usefixtures("feature_smmp_present")
```
Comment (saimidu) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smtrcomp.py:191]: Please leave 2 lines of space before the start of a function.

```suggestion

    @pytest.mark.usefixtures("feature_smmp_present")
```

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smtrcomp.py [hunk 17]
```diff
@@ -217,7 +217,7 @@ def test_smmp(self, sagemaker_session, ecr_image, framework_version, efa_instanc
...
```
Comment (saimidu) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smtrcomp.py:220]: Please leave 2 lines of space before the start of a function.

```suggestion

    @pytest.mark.usefixtures("feature_smmp_present")
```

File: test/test_utils/imageutils.py (hunk 5 above)
Comment (kenny-ezirim) [reply test/test_utils/imageutils.py:5]: It is meant to factor out all helper functions used for image. In the future, we can refactor the code to use imageutil module and remove helper functions that appear in other places.

File: test/test_utils/imageutils.py (hunk 7 above)
Comment (kenny-ezirim) [reply test/test_utils/imageutils.py:20]: Same here as above.

File: test/test_utils/imageutils.py (hunk 6 above)
Comment (kenny-ezirim) [reply test/test_utils/imageutils.py:16]: Same here as above.

File: test/test_utils/imageutils.py (hunk 8 above)
Comment (kenny-ezirim) [reply test/test_utils/imageutils.py:31]: Same here as above.

File: test/test_utils/imageutils.py (hunk 9 above)
Comment (kenny-ezirim) [reply test/test_utils/imageutils.py:48]: Same here as above.

File: test/dlc_tests/conftest.py [hunk 18]
```diff
@@ -811,6 +843,23 @@ def pytest_generate_tests(metafunc):
...
+            # iterate through image candidates and select images with labels that match all nightly fixture labels
+            for image_candidate in images_to_parametrize:
+                if all([are_valid_fixture_labels_present(image_candidate, nightly_labels) for _, nightly_labels in func_nightly_fixtures.items()]):
+                    nightly_images_to_parametrize.append(image_candidate)
+            """
+            print(f"Function: {metafunc.function.__name__}")
+            print(f"Nightly Fixture Labels: {func_nightly_fixtures}")
+            print(f"Number of images to parametrize: {len(images_to_parametrize)}")
+            print(f"Number of nightly images to parametrize: {len(nightly_images_to_parametrize)}\n")
+            """
```
Comment (saimidu) [inline test/dlc_tests/conftest.py:859]: Can this be removed in the final version?

File: test/dlc_tests/sanity/test_nightly_fixtures.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
+import pytest
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_nightly_fixtures.py:1]: LGTM. Let's remove this test script.

File: test/dlc_tests/test_type_images.json [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
```
Comment (saimidu) [inline test/dlc_tests/test_type_images.json:1]: LGTM. Let's remove this file for the approval.

File: testspec.yml [hunk 21]
```diff
@@ -23,6 +23,7 @@ phases:
       - pip install scheduler/.
       - echo Running pytest $TEST_TYPE tests on $DLC_IMAGES...
       - export PYTHONPATH=$PYTHONPATH:$(pwd)/src
+      - export CODEBUILD_SRC_DIR_DLC_IMAGES_JSON=$(pwd)/test/dlc_tests
```
Comment (saimidu) [inline testspec.yml:26]: Let's remove this for the final approval.

File: pytorch/buildspec.yml [hunk 22]
```diff
@@ -46,36 +46,36 @@ context:
...
-  #   tag_python_version: &TAG_PYTHON_VERSION py38
-  #   cuda_version: &CUDA_VERSION cu113
-  #   os_version: &OS_VERSION ubuntu20.04
-  #   tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION, "-e3" ]
-  #   docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /, *CUDA_VERSION, /Dockerfile.,
-  #                        *DEVICE_TYPE ]
-  #   target: e3
-  #   context:
-  #     <<: *TRAINING_CONTEXT
+  BuildE3CPUPTTrainPy3DockerImage:
```
Comment (saimidu) [inline pytorch/buildspec.yml:49]: Please revert all the changes that need to be reverted in this file for the final approval.

**Review by saimidu — DISMISSED** [review]
Approved.

**Comment by saimidu** [comment]
Test parametrization works as intended on commit [859f0c5](https://github.com/aws/deep-learning-containers/pull/2090/commits/859f0c583f6296bafdfe3188e1188d6dbb4cf75b).

File: test/dlc_tests/conftest.py [hunk 23]
```diff
@@ -124,8 +125,9 @@ def feature_aws_framework_present():
 
 def pytest_addoption(parser):
     default_images = test_utils.get_dlc_images()
+    images = default_images.split(" ") if default_images else []
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/dlc_tests/conftest.py:128]: I encountered the case where default_images was None while testing. I just think we should check regardless, since the get_dlc_images does not guarantee not having a NoneType.

**Review by saimidu — APPROVED** [review]

**Review by Lokiiiiii — APPROVED** [review]
